### ISSUE TITLE
RequireThisCheck: pattern variables produce false positives

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#10969]
Identified at https://github.com/checkstyle/checkstyle/issues/7504#issuecomment-974776332 ,

````
$ cat TestClass.java
public class Test {
String p, s, n;

public static void main(String... arguments) {
new Test().run(arguments);
}

public void run(String... arguments) {
this.p = null;
this.s = null;
this.n = null;

Object o = 45;
if (!(o instanceof String p && p.equals("sd")) || !p.equals("wq")) {
p = "a"; // instance var p
} else if (!(o instanceof Integer s && o instanceof String n)) {
p = "b"; // local var p
s = "b"; // instance var s
n = "b"; // instance var n
} else {
p = "c"; // local var p
s = 41; // local var s
n = "c"; // local var n
}


System.out.println(this.p + ":" + this.s + ":" + this.n);
}
}


$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="RequireThis">
  <property name="validateOnlyOverlapping" value="false"/>
</module>
    </module>
</module>

$ java -jar checkstyle-9.0.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] TestClass.java:15:43: Reference to instance variable 'p' needs "this.". [RequireThis]
[ERROR] TestClass.java:15:48: Reference to instance variable 'p' needs "this.". [RequireThis]
[ERROR] TestClass.java:15:68: Reference to instance variable 'p' needs "this.". [RequireThis]
[ERROR] TestClass.java:16:25: Reference to instance variable 'p' needs "this.". [RequireThis]
[ERROR] TestClass.java:17:51: Reference to instance variable 's' needs "this.". [RequireThis]
[ERROR] TestClass.java:17:76: Reference to instance variable 'n' needs "this.". [RequireThis]
[ERROR] TestClass.java:18:25: Reference to instance variable 'p' needs "this.". [RequireThis]
[ERROR] TestClass.java:19:25: Reference to instance variable 's' needs "this.". [RequireThis]
[ERROR] TestClass.java:20:25: Reference to instance variable 'n' needs "this.". [RequireThis]
[ERROR] TestClass.java:22:25: Reference to instance variable 'p' needs "this.". [RequireThis]
[ERROR] TestClass.java:23:25: Reference to instance variable 's' needs "this.". [RequireThis]
[ERROR] TestClass.java:24:25: Reference to instance variable 'n' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 12 errors.
````

As the comments demonstrate, only really lines 16, 19, and 20 should have violations as they are accessing the instance variables.

You can confirm this by running the code, which as is outputs `a:null:null`.

Thanks @nmancus1 for explaining all this to me.

**Comment by Vyom-Yadav** [comment]
On it.

**Comment by nmancus1** [comment]
@Vyom-Yadav it might be a good idea to discuss high level implementation of expression evaluation idea in here first before sending PR.

**Comment by Vyom-Yadav** [comment]
@nmancus1 Of course, the plan is to simplify the boolean expression, and check whether the pattern variable has been introduced or not. Before moving to complicated examples, I will quickly explain what we need to do in basic examples.

 In all examples, there is an instance var `p` present.

### Example 1-
```
Object o = "ds";
if(o instanceof String p || p.equals("a")) { // second 'p' the the instance var.
}
```
Pattern var has not been introduced since it is used with `||`.
**Q1.** How will this be implemented in code-
We will keep increasing the scope of pattern var until we hit `||` as its ancestor, this shouldn't be a problem, I implemented this earlier before finding more complications involved.

Some more examples of the same flavor-
```
if(o instanceof String p 
                && p.equals("a") || p.equals("a")) { // 1st p is pattern var, 2nd is instance var
}
```
In the above example also pattern var has not been introduced as we encountered an `||` condition. But in this condition, the scope of the pattern var (in terms of line no. and column no.) would be increased and will be present until before `||`.

You get the point of how this works, we will `break` from the loop when we hit `||` as an ancestor.

### Example 2-
These are the types of examples where the pattern var has been fully introduced.
```
if(o instanceof String p
                && p.equals("a") && p.equals("a")) {
            // pattern var p is accessible here
}
```
We didn't encounter any `||` so pattern var is present in the scope of the if statement, again this will be implemented using the logic mentioned in example 1 with a rule that if we hit `TokenTypes.EXPR` without catching an `||` then update its scope to the full condition.

### The real fun starts when we have to deal with `!`.
So the plan to deal with the undermentioned examples would be to simplify the expression and get rid of `!`.
Now if we have an even number of `!` as ancestors of pattern var then we can simplify the expression by merging the even num of `!`

#### Example-
```
if(!(!(o instanceof String p) || p.equals("a"))) {
}
```
can be simplified to
```
if(((o instanceof String p) && !p.equals("a"))) {
    // pattern var p is accessible here.
}
```
Q1. How will this be implemented in code-
Let's look at the parsed tree of both the codes-
`!(!(o instanceof String p) || p.equals("a"))`
```
|--EXPR -> EXPR [69:11]
    `--LNOT -> ! [69:11]
        |--LPAREN -> ( [69:12]
        |--LOR -> || [69:38]
        |   |--LNOT -> ! [69:13]
        |   |   |--LPAREN -> ( [69:14]
        |   |   |--LITERAL_INSTANCEOF -> instanceof [69:17]
        |   |   |   |--IDENT -> o [69:15]
        |   |   |   `--PATTERN_VARIABLE_DEF -> PATTERN_VARIABLE_DEF [69:28]
        |   |   |       |--MODIFIERS -> MODIFIERS [69:28]
        |   |   |       |--TYPE -> TYPE [69:28]
        |   |   |       |   `--IDENT -> String [69:28]
        |   |   |       `--IDENT -> p [69:35]
        |   |   `--RPAREN -> ) [69:36]
        |   `--METHOD_CALL -> ( [69:49]
        |       |--DOT -> . [69:42]
        |       |   |--IDENT -> p [69:41]
        |       |   `--IDENT -> equals [69:43]
        |       |--ELIST -> ELIST [69:50]
        |       |   `--EXPR -> EXPR [69:50]
        |       |       `--STRING_LITERAL -> "a" [69:50]
        |       `--RPAREN -> ) [69:53]
        `--RPAREN -> ) [69:54]
```
`((o instanceof String p) && !p.equals("a"))`
```
|--EXPR -> EXPR [69:11]
    |--LPAREN -> ( [69:11]
    |--LAND -> && [69:36]
    |   |--LPAREN -> ( [69:12]
    |   |--LITERAL_INSTANCEOF -> instanceof [69:15]
    |   |   |--IDENT -> o [69:13]
    |   |   `--PATTERN_VARIABLE_DEF -> PATTERN_VARIABLE_DEF [69:26]
    |   |       |--MODIFIERS -> MODIFIERS [69:26]
    |   |       |--TYPE -> TYPE [69:26]
    |   |       |   `--IDENT -> String [69:26]
    |   |       `--IDENT -> p [69:33]
    |   |--RPAREN -> ) [69:34]
    |   `--LNOT -> ! [69:39]
    |       `--METHOD_CALL -> ( [69:48]
    |           |--DOT -> . [69:41]
    |           |   |--IDENT -> p [69:40]
    |           |   `--IDENT -> equals [69:42]
    |           |--ELIST -> ELIST [69:49]
    |           |   `--EXPR -> EXPR [69:49]
    |           |       `--STRING_LITERAL -> "a" [69:49]
    |           `--RPAREN -> ) [69:52]
    `--RPAREN -> ) [69:53]
```
Now we can start from the topmost `!`,  and start negating the `||` and `&&` present, remove `!` which is parent of `LITERAL_INSTANCEOF` and add `!` node as a parent of `METHOD_CALL` and after doing all that, remove the topmost `!` token and modify the tree.
Also if we encounter another `!` which is not parent of `LITERAL_INSTANCEOF`, we will repeat the same algorithm for it.
We would have to prepend `!` in every place that is not already negated (in case it's already negated, they will cancel out).
### One thing we have to keep in mind during negating `||` and `&&` is operator precedence
Example of operator precedence-
```
if(!(!(o instanceof String p) || p.equals("ds") && p.equals("das"))) {
}
```
will simplify to -
```
if(((o instanceof String p) && !(p.equals("sd") && p.equals("das")))) {
}
```
and not to-
```
if(((o instanceof String p) && !p.equals("sd") || !p.equals("das"))) {
}
```
To solve this problem we can just add `LPAREN` and `RPAREN` around `&&` statements

And after doing mentioned above we will get the expression `((o instanceof String p) && !p.equals("a"))` which we will check using rules we developed in earlier sections.

#### Now what if we don't have an even number of `!` as the ancestor of pattern var then we would have to add a `!` outside the whole expression and then process it using the rules for even `!`. 

Example-
```
if (!(o instanceof String p) || p.equals("a")) { // ATTENTION: both p are pattern var
}
```
we will prepend `!` to check whether it will be in scope in the upcoming `else` and `else if` statements and in the 
[main block if an exception is thrown](https://docs.oracle.com/javase/specs/jls/se17/html/jls-6.html#jls-6.3.2)

### Fun++;
Now we are coming close to our end as we have to evaluate 2 or more pattern variables introduces in one expression
#### Example-
```
String s, p; // instance variables 

if (!(o instanceof String s) || s.equals("a") && !(o instanceof Integer p) || p.equals("ad")) {
            System.out.println(s); // instance var 's'
            System.out.println(p); // instance var 'p'
}
else {
            System.out.println(s); // pattern var 's'
            System.out.println(p); // instance var 'p'
}
```
In the example above both pattern var have only one `!` as their ancestor so we would have to negate them to check whether they would exist in `else` block or not, this would be negated as-
```
!(o instanceof String s) || s.equals("a") && !(o instanceof Integer p) || p.equals("ad")
                                           |
                                           |
                                           V
(o instanceof String s) && !(s.equals("a") && !(o instanceof Integer p)) && !p.equals("ad")
```
From here we can conclude that `s` was successfully introduced in `else` block but `p` now has two `!` as its ancestor so we would have to simplify it which would be done as-
```
(o instanceof String s) && !(s.equals("a") && !(o instanceof Integer p)) && !p.equals("ad")
                                           |
                                           |
                                           V
(o instanceof String s) && (!s.equals("a") || (o instanceof Integer p)) && !p.equals("ad")
```
Now as `p` has `||` as its ancestor, it is not introduced.

Now I couldn't find a case where one pattern var is introduced in `if` block and the other is introduced in `else` block which I feel isn't possible, consider the example-
```
if (!(o instanceof String s) && o instanceof String p) {
            System.out.println(s); / instance var
            System.out.println(p); // pattern var
}
else {
            System.out.println(s); / instance var
            System.out.println(p); // instance var
}
```
Now with whichever combination you try, it isn't possible that one is introduced in `if` and the other is introduced in `else` though both can be introduced in `if` or `else`. 

So now concluding my summary with a few points-

- We have to check for scope outside `if` iff pattern var has odd `!` as its ancestor
- If it is present in `else` then it would be present in `elif` also
- We have to add parentheses around `&&` statements before negating
- We have to also change `==` to `!=`, etc.
- We will reduce 3 `!` (ancestor of pattern var) to 1 and 4 to 2 (you get the point)

Tried to cover all cases, naturally can miss some, just ping me if you have any doubts or queries regarding anything I mentioned above.

Pinging other members to get their opinion-
@romani @rnveach @pbludov @strkkk

**Comment by nmancus1** [comment]
> and in the main block if an exception is thrown

I was not aware of this nuance; this makes things a bit more complex.

I can't help but wonder if we should implement this pattern variable "expression evaluation" in a simpler check first, such as the new check proposed in https://github.com/checkstyle/checkstyle/issues/10970#issuecomment-974900102 (not approved yet). `RequireThis` is a fairly complex check, and it would be good to see your plan working in a simpler check before coupling the challenges that expression evaluation brings with integration of pattern variables into `RequireThis`.

**Comment by nmancus1** [comment]
@Vyom-Yadav how do you plan to store the state of the AST/ expression as it is evaluated?

**Comment by Vyom-Yadav** [comment]
@nmancus1 
> I was not aware of this nuance; this makes things a bit more complex.

[This](https://docs.oracle.com/en/java/javase/17/language/pattern-matching-switch-expressions-and-statements.html) is also a new feature, although a preview feature, may become permanent in the future.

>I can't help but wonder if we should implement this pattern variable "expression evaluation" in a simpler check first, su...

I also feel the same, resolving the scope of pattern variables can be used in many checks eg - https://github.com/checkstyle/checkstyle/pull/10930 (though a PR right now), `RequireThis` and in many checks that might be added in the future.

>how do you plan to store the state of the AST/ expression as it is evaluated?

We will have to create a new `EXPR` node inside the check itself, keep adding children to it as we evaluate the original `EXPR` node (basically copy children of the original `EXPR` node and make changes to the copied ones), then check the new `EXPR` to meet the criteria and at the end mark the scope of pattern var which is present under the original `EXPR` node.

**Comment by nmancus1** [comment]
@Vyom-Yadav just to be very clear, there can be three results of expression evaluation, right?

1. Given pattern variable is in scope of block (conditional or loop) guarded/ controlled by expression
2. Given pattern variable in in scope of following `else if` and `else` blocks in case of conditionals
3. Given pattern variable is in scope of block enclosing conditional statements/loops mentioned above (outer block)

> We will have to create a new EXPR node inside the check itself

We discourage the instantiation of `DetailAst` outside of parser, and to have all `DetailAst` methods available for this expression evaluation seems to be extraneous anyway. It is probably better to create a new, simpler AST like structure for expression evaluation.

**Comment by Vyom-Yadav** [comment]
@nmancus1 
> just to be very clear, there can be three results of expression evaluation, right?

Yes, those are the possible scopes **without** including preview features, also I missed one example-
```
boolean a() {
        Object o = "as";
        if(!(o instanceof String s)) {
            return false;
        }
        return s.equals("ad"); // pattern var 's'
    }
```
> We discourage the instantiation of DetailAst outside of parser.......

Understood, shouldn't be a problem using some sort of tree to to do it.
